packages/core/src/api/apis/definitions/oauthrequest.ts

```diff
@@ -0,0 +1,173 @@
+  | string[] /** Array of individual scope strings */
+  | OAuthScopes;
+
+/**
+ * Options used to open a login popup.
+ */
+export type LoginPopupOptions = {
+  /**
+   * The URL that the auth popup should point to
+   */
+  url: string;
+
+  /**
+   * The name of the popup, as in second argument to window.open
+   */
+  name: string;
+
+  /**
+   * The origin of the final popup page that will post a message to this window.
+   */
+  origin: string;
+
+  /**
+   * The width of the popup in pixels, defaults to 500
+   */
+  width?: number;
+
+  /**
+   * The height of the popup in pixels, defaults to 700
+   */
+  height?: number;
+};
+
+/**
+ * Information about the auth provider that we're requesting a login towards.
+ *
+ * This should be show to the user so that they can be informed about what login is being requested
+ * before a popup is shown.
+ */
+export type AuthProviderInfo = {
+  /**
+   * Title for the auth provider, for example "GitHub"
+   */
+  title: string;
+
+  /**
+   * Icon for the auth provider.
+   */
+  icon: IconComponent;
+};
+
+/**
+ * Describes how to handle auth requests. Both how to show them to the user, and what to do when
+ * the user accesses the auth request.
+ */
+export type AuthRequesterOptions<T> = {
+  /**
+   * Information about the auth provider, which will be forwarded to auth requests.
+   */
+  info: AuthProviderInfo;
+
+  /**
+   * Implementation of the auth flow, which will be called synchronously when
+   * triggerAuth() is called on an auth requests.
+   */
+  authHandler(scope: OAuthScopes): Promise<T>;
+};
+
+/**
+ * Function used to trigger new auth requests for a set of scopes.
+ *
+ * The returned promise will resolve to the same value returned by the authHandler in the
+ * AuthRequesterOptions. Or rejected, if the request is rejected.
+ *
+ * This function can be called multiple times before the promise resolves. All calls
+ * will be merged into one request, and the scopes forwarded to the authHandler will be the
+ * union of all requested scopes.
+ */
+export type AuthRequester<T> = (scope: OAuthScopes) => Promise<T>;
+
+/**
+ * An auth request for a single auth provider.
+ */
+export type AuthRequest = {
+  /**
+   * Information about the auth provider, as given in the AuthRequesterOptions
+   */
+  info: AuthProviderInfo;
+
+  /**
+   * Rejects the request, causing all pending AuthRequester calls to fail with "RejectedError".
+   */
+  reject: () => void;
+
+  /**
+   * Synchronously calls the auth handler with all scope currently in the request.
+   */
+  triggerAuth(): Promise<void>;
+};
+
+/**
+ * Provides helpers for implemented OAuth login flows within Backstage.
+ */
+export type OAuthRequestApi = {
+  /**
+   * Show a popup pointing to a URL that starts an OAuth flow.
+   *
+   * The redirect handler of the flow should use postMessage to communicate back
+   * to the app window.
+   */
+  showLoginPopup(options: LoginPopupOptions): Promise<any>;
```

Comment:
Unclear what the return type is. Document or narrow?

Reply:
Added a note about what is returned, I think we might need to revisit this though.